_removeDataStoreWithIdentifier should fail when the data store is being used by web page

szewai opened the pull request:
#### 1c291a95fe759ba4fa23b4ce7a50698b5ade2e0f
<pre>
_removeDataStoreWithIdentifier should fail when the data store is being used by web page
<a href="https://bugs.webkit.org/show_bug.cgi?id=253555">https://bugs.webkit.org/show_bug.cgi?id=253555</a>
rdar://106407998

Reviewed by Chris Dumez.

To avoid undefined behaviors and unexpected results, we don&apos;t allow data store to be removed while it is being used by
web page.

API Test: WKWebsiteDataStore.RemoveDataStoreWithIdentifierErrorWhenInUse

* Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
(WebKit::WebsiteDataStore::removeDataStoreWithIdentifier):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::existingDataStoreForIdentifier):
(WebKit::WebsiteDataStore::hasActivePages):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebsiteDatastore.mm:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/261399@main">https://commits.webkit.org/261399@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fae1a5a05f1d7a24e6be5118d9c026a658f8a467

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/111573 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/20707 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/176 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/3333 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/120331 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/115632 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/22063 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/11791 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/86/builds/3092 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/117337 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/16385 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/99533 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/104352 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/98341 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/111 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/45145 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/13194 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/108 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/86958 "13 api tests failed or timed out") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/13688 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/9554 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/19133 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/52092 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7939 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/15671 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
```diff
@@ -282,11 +282,23 @@ static String defaultWebsiteDataStoreRootDirectory()
+                stringBuilder.append(page.toUInt64());
+                stringBuilder.append(',');
+            }
+            RELEASE_LOG_ERROR(Storage, "WebsiteDataStore::removeDataStoreWithIdentifier data store is used by web pages { %s }", stringBuilder.toString().utf8().data());
```
Is the logging of page ids actually that useful? We're making code a bit less efficient and a bit more complicated just for this logging so I want to make sure having those ids is actually useful for debugging purposes.

szewai replied:
The loggings aims to help client find out why they cannot remove the store (e.g. they might accidentally hold strong ref to a web view or the web view is in an autorelease pool.). I think we log the page ids in other places, so with this new log they may be able to figure out which view is problematic.

I am not sure how useful the logging would be in real world use cases though; I could remove it if you think it's unnecessary

cdumez commented on Source/WebKit/UIProcess/WebProcessPool.cpp:
```diff
@@ -1190,11 +1190,11 @@ void WebProcessPool::pageEndUsingWebsiteDataStore(WebPageProxyIdentifier pageID,
 }
 
-bool WebProcessPool::hasPagesUsingWebsiteDataStore(WebsiteDataStore& dataStore) const
+HashSet<WebPageProxyIdentifier> WebProcessPool::pagesUsingWebsiteDataStore(WebsiteDataStore& dataStore) const
```
May be nice to return a `const HashSet<WebPageProxyIdentifier>*` to avoid copying.

cdumez replied:
That's if we *really* need the page IDs.

szewai replied:
Page ids removed.

cdumez commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
```diff
@@ -2352,6 +2362,17 @@ void WebsiteDataStore::resumeDownload(const DownloadProxy& downloadProxy, const
+{
+    HashSet<WebPageProxyIdentifier> result;
+    for (auto& pool : WebProcessPool::allProcessPools()) {
+        for (auto& page : pool->pagesUsingWebsiteDataStore(*this))
```
We can probably leverage this Hashset function to avoid doing our own loop:
```
template<typename IteratorType>
bool add(IteratorType begin, IteratorType end);
```

szewai replied:
This is removed.

cdumez commented on Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
```diff
@@ -477,6 +478,7 @@ class WebsiteDataStore : public API::ObjectImpl<API::Object::Type::WebsiteDataSt
     static void removeMediaKeys(const String& mediaKeysStorageDirectory, const HashSet<WebCore::SecurityOriginData>&);
 
     void registerWithSessionIDMap();
+    HashSet<WebPageProxyIdentifier> activePages();
```
Seems like this should be const.

szewai replied:
Updated.

cdumez reviewed (approved):
LGTM

szewai committed:
_removeDataStoreWithIdentifier should fail when the data store is being used by web page
https://bugs.webkit.org/show_bug.cgi?id=253555
rdar://106407998

Reviewed by Chris Dumez.

To avoid undefined behaviors and unexpected results, we don't allow data store to be removed while it is being used by
web page.

API Test: WKWebsiteDataStore.RemoveDataStoreWithIdentifierErrorWhenInUse

* Source/WebKit/UIProcess/WebsiteData/Cocoa/WebsiteDataStoreCocoa.mm:
(WebKit::WebsiteDataStore::removeDataStoreWithIdentifier):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.cpp:
(WebKit::WebsiteDataStore::existingDataStoreForIdentifier):
(WebKit::WebsiteDataStore::hasActivePages):
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStore.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebsiteDatastore.mm:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/261399@main
